// wangeditor5-for-vue2/dist/es/core/editor.js
import { WeEditable } from './editable';
import Vue from 'vue';
import { WeToolbar } from './toolbar';
export const WeEditor = Vue.extend({
    name: 'WeEditor',
    props: {
        toolbarOption: {
            type: Object,
            required: true,
        },
        toolbarClass: [String, Object, Array],
        toolbarStyle: [String, Object],
        toolbarReloadbefore: {
            type: Function,
            default: () => () => { },
        },
        editableOption: {
            type: Object,
            required: true,
        },
        editableClass: [String, Object],
        editableStyle: [String, Object],
        editableReloadbefore: {
            type: Function,
            default: () => () => { },
        },
        json: String,
        html: String,
    },
    render(h) {
        return h('div', [
            h(WeToolbar, {
                class: this.toolbarClass,
                style: this.toolbarStyle,
                props: {
                    option: this.toolbarOption,
                    reloadbefore: this.toolbarReloadbefore,
                },
            }),
            h(WeEditable, {
                class: this.editableClass,
                style: this.editableStyle,
                props: {
                    option: this.editableOption,
                    json: this.json,
                    html: this.html,
                    reloadbefore: this.editableReloadbefore,
                },
                on: {
                    'update:json': (value) => this.$emit('update:json', value),
                    'update:html': (value) => this.$emit('update:html', value),
                },
            }),
        ]);
    },
});
WeEditor.install = function (vue) {
    vue.component('WeEditor', WeEditor);
};
